Fix-#90: Check whether the environment passed to "-e" option does exist

glammers1 opened the pull request:
## Description
Closes #90 

## Changes
**UPDATE. Please make sure you read this message as well https://github.com/boostercloud/booster/pull/95#issuecomment-633735766**

- I added a new variable in the `BoosterConfig` class called `configuredEnvironments`. That variable aims to have all the environments that the user is supposed to set in their `config.ts` file by calling the `Booster.configure` function. Every time the function is called, a new value is included into the array. 
- Accessing a provider is possible as long as the user has configured at least one environment along with its provider. Thus, before checking if a provider was set, we check whether the environment used to execute an action (deploy, nuke...) exists and match with some of the available.

Some output examples:

- The environment provided to execute the action doesn't match with any of the available:

```
nuke -e development -f
ℹ boost nuke [development] 🧨
✖ Error: The environment 'development' does not match with any of the environments you used to configure your Booster project which are: 'production, stage'
    at BoosterConfig.get provider [as provider] (/Users/glammers/development/theam/booster/packages/framework-types/dist/config.js:79:19)
    at exports.nukeCloudProviderResources (/Users/glammers/development/theam/booster/packages/cli/src/services/provider-service.ts:48:17)
    at /Users/glammers/development/theam/booster/packages/cli/src/commands/nuke.ts:20:9
    at /Users/glammers/development/theam/booster/packages/cli/src/common/script.ts:53:23
    at /Users/glammers/development/theam/booster/node_modules/fp-ts/lib/TaskEither.js:127:33
    at /Users/glammers/development/theam/booster/node_modules/fp-ts/lib/Task.js:114:95
    at Script.done (/Users/glammers/development/theam/booster/packages/cli/src/common/script.ts:102:22)
    at Nuke.run (/Users/glammers/development/theam/booster/packages/cli/src/commands/nuke.ts:69:5)
    at Nuke._run (/Users/glammers/development/theam/booster/node_modules/@oclif/command/lib/command.js:42:20)
    at Config.runCommand (/Users/glammers/development/theam/booster/node_modules/@oclif/config/lib/config.js:172:9)
```

- There is not any configured environment:

```
nuke -e development -f
ℹ boost nuke [development] 🧨
✖ Error: You haven't configured any environment. It is required to specify the deployment environments in 'src/config/config.ts'
    at BoosterConfig.get provider [as provider] (/Users/glammers/development/theam/booster/packages/framework-types/dist/config.js:79:19)
    at exports.nukeCloudProviderResources (/Users/glammers/development/theam/booster/packages/cli/src/services/provider-service.ts:48:17)
    at /Users/glammers/development/theam/booster/packages/cli/src/commands/nuke.ts:20:9
    at /Users/glammers/development/theam/booster/packages/cli/src/common/script.ts:53:23
    at /Users/glammers/development/theam/booster/node_modules/fp-ts/lib/TaskEither.js:127:33
    at /Users/glammers/development/theam/booster/node_modules/fp-ts/lib/Task.js:114:95
    at Script.done (/Users/glammers/development/theam/booster/packages/cli/src/common/script.ts:102:22)
    at Nuke.run (/Users/glammers/development/theam/booster/packages/cli/src/commands/nuke.ts:69:5)
    at Nuke._run (/Users/glammers/development/theam/booster/node_modules/@oclif/command/lib/command.js:42:20)
    at Config.runCommand (/Users/glammers/development/theam/booster/node_modules/@oclif/config/lib/config.js:172:9)
```

## Checks
- [X] Project Builds
- [X] Project passes tests and checks
- [ ] Updated documentation accordingly at [the docs repo](https://github.com/boostercloud/docs)
- [ ] Added a link pointing to the documentation PR in this PR

## Additional information

- I am not sure if this is the best place to check whether or not the environment exists.
- I was about to create somewhat a `BoosterConfigTest` class for the test to extend the current `BoosterConfig` so that we don't need to do this `config.configuredEnvironments.push('test')` all the time. Let me know if it makes sense.

glammers1 commented:
Disregard PR's description. I just added several commits to change the approach. I added the validation to `readProjectConfig` method in the `config-server.ts` file. This way, we just check if the environment exists when executing a `nuke`, `deploy`, or `run` rather than every time we get the `provider`. I think it makes more sense. Let me know if you like this new approach. If so, there are no tests for the file `config-server.ts`, so I would add them since I've added some logic on it.

tainguyenbui commented on packages/cli/src/services/config-service.ts:
```diff
@@ -28,7 +28,18 @@ function readProjectConfig(userProjectPath: string): Promise<BoosterConfig> {
+function checkEnvironmentWasConfigured(config: BoosterConfig): void {
+  if (!config.configuredEnvironments.includes(config.environmentName)) {
+    const errorMessage = config.configuredEnvironments.length
+      ? `The environment '${config.environmentName}' does not match with any of the environments` +
```
```suggestion
      ? `The environment '${config.environmentName}' does not match any of the environments` +
```

tainguyenbui commented on packages/cli/src/services/config-service.ts:
```diff
@@ -28,7 +28,18 @@ function readProjectConfig(userProjectPath: string): Promise<BoosterConfig> {
+  if (!config.configuredEnvironments.includes(config.environmentName)) {
+    const errorMessage = config.configuredEnvironments.length
+      ? `The environment '${config.environmentName}' does not match with any of the environments` +
+        ` you used to configure your Booster project which are: '${config.configuredEnvironments.join(', ')}'`
```
```suggestion
        ` you used to configure your Booster project, which are: '${config.configuredEnvironments.join(', ')}'`
```

alvaroloes commented on packages/framework-types/src/config.ts:
```diff
@@ -104,6 +105,12 @@ export class BoosterConfig {
+  public addConfiguredEnvironment(environmentName: string): void {
+    if (!this.configuredEnvironments.includes(environmentName)) {
+      this.configuredEnvironments.push(environmentName)
+    }
```
You can use a `Set<string>` instead of an array for the `configuredEnvironments` property and then you can omit the `if` 😃 (as Sets can't contain duplicates)
```typescript
this.configuredEnvironments.add(environmentName)
```

glammers1 replied:
Done! Thank you for pointing this out.

alvaroloes commented on packages/cli/src/services/config-service.ts:
```diff
@@ -28,7 +28,18 @@ function readProjectConfig(userProjectPath: string): Promise<BoosterConfig> {
+    const errorMessage = config.configuredEnvironments.length
+      ? `The environment '${config.environmentName}' does not match with any of the environments` +
+        ` you used to configure your Booster project which are: '${config.configuredEnvironments.join(', ')}'`
+      : "You haven't configured any environment. It is required to specify the deployment environments in 'src/config/config.ts'"
```
Take into account that the file where the user puts its config can be anywhere and have any name (it is common to have several files, one per environment, like `config/production.ts` and `config/local.ts`). So I would change this message to something like:
```suggestion
      : "You haven't configured any environment. Please make sure you have at least one environment configured by calling `Booster.configure` method (normally done inside the folder 'src/config')"
```

alvaroloes commented:
bot: integration

alvaroloes commented:
bot: integration

alvaroloes commented:
bot: integration

alvaroloes commented:
bot: integration

alvaroloes commented:
bot: integration

alvaroloes commented:
bot: integration

alvaroloes commented:
bot: integration

alvaroloes commented:
bot: integration

alvaroloes commented:
bot: integration

alvaroloes commented:
bot: integration

alvaroloes commented:
bot: integration

alvaroloes commented:
bot: integration

alvaroloes commented:
bot: integration

alvaroloes commented:
bot: integration

alvaroloes commented:
Please ignore comments like `bot: integraion`. I'm doing tests ;-)

alvaroloes commented:
bot: integration

glammers1 commented:
I've just added a test file for the `config-service` since I had added a new function there, and that file didn't have its specs. I was struggling with mocking the global `require` function, and I finally came up with a solution. Apart from that, the tests are straightforward. If you agree with the implementation, this PR is ready to be merged. [Tai's comment ](https://github.com/boostercloud/booster/pull/95#pullrequestreview-418071686) is out of the scope of this PR, and it should be addressed in another one.

@alvaroloes `create integration test check` failed for some reason, could you take a look at it?

glammers1 commented:
bot: integration

alvaroloes commented:
The integration check only works for PRs created from now on. So you can ignore that check in this PR 😉

alvaroloes commented on packages/cli/test/services/config-service.test.ts:
```diff
@@ -0,0 +1,77 @@
+      config.addConfiguredEnvironment('test')
+      Module.prototype.require = function() {
+        return new BoosterAppMock(config)
+      }
```
💥
